File: events/2015/07/solution.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"reflect"
	"strconv"
	"strings"
	"time"
	"yod_aoc/utils/go/aoc_utils"
)

func update_connections(connections map[string]interface{}) map[string]interface{} {
	for wire, command := range connections {
		if reflect.TypeOf(command).String() == "uint16" {
			continue
		}

		if len(command.([]interface{})) == 1 {
			new_command, ok := command.([]interface{})[0].(string)

			if ok {
				num, err := strconv.ParseUint(new_command, 10, 16)
				if err == nil {
					connections[wire] = uint16(num)
				}
			}
		}
	}

	return connections
}

func solve_connection(connections map[string]interface{}, wire string) (uint16, map[string]interface{}) {
	connections = update_connections(connections)
	wc, exists := connections[wire]

	if !exists {
		return 0, connections
	}

	wcType := reflect.TypeOf(wc).String()

	if wcType == "uint16" {
		return wc.(uint16), connections
	}

	length := len(wc.([]interface{}))

	if reflect.TypeOf(wc).String() == "uint16" {
		return wc.(uint16), connections
	}

	if length == 1 {

		val := wc.([]interface{})[0]

		if num, ok := val.(uint16); ok {
			connections[wire] = num
			return num, connections
		}

		if str, ok := val.(string); ok {
			if num, err := strconv.ParseUint(str, 10, 16); err == nil {
				result := uint16(num)
				connections[wire] = result
				return result, connections
			}

			result, connections := solve_connection(connections, str)
			connections[wire] = result
			return result, connections
		}
	} else if length == 2 {
		nv := wc.([]interface{})[1]

		nv_num, ok := nv.(uint16)
		if !ok {
			nvStr, isStr := nv.(string)
			if isStr {
				num, err := strconv.ParseUint(nvStr, 10, 16)
				if err == nil {
					nv_num = uint16(num)
					ok = true
				}
			}

			if !ok {
				nv_num, connections = solve_connection(connections, nv.(string))
			}
		}

		ans := uint16(^nv_num)
		connections[wire] = ans
		return ans, connections
	} else if length == 3 {
		op := wc.([]interface{})[1].(string)

		a := wc.([]interface{})[0]
		b := wc.([]interface{})[2]

		a_num, a_ok := a.(uint16)
		b_num, b_ok := b.(uint16)

		if !a_ok {
			aStr, isStr := a.(string)

			if isStr {
				num, err := strconv.ParseUint(aStr, 10, 16)
				if err == nil {
					a_num = uint16(num)
					a_ok = true
				}
			}

			if !a_ok {
				a_num, connections = solve_connection(connections, a.(string))
			}
		} else {
			a = a_num
		}

		if !b_ok {
			bStr, isStr := b.(string)

			if isStr {
				num, err := strconv.ParseUint(bStr, 10, 16)
				if err == nil {
					b_num = uint16(num)
					b_ok = true
				}
			}

			if !b_ok {
				b_num, connections = solve_connection(connections, b.(string))
			}
		} else {
			b = b_num
		}

		if op == "AND" {
			ans := uint16(a_num & b_num)
			connections[wire] = ans
			return ans, connections
		} else if op == "OR" {
			ans := uint16(a_num | b_num)
			connections[wire] = ans
			return ans, connections
		} else if op == "LSHIFT" {
			ans := uint16(a_num << b_num)
			connections[wire] = ans
			return ans, connections
		} else if op == "RSHIFT" {
			ans := uint16(a_num >> b_num)
			connections[wire] = ans
			return ans, connections
		}
	}

	return 0, connections
}

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
func part1(input []string) uint16 {
	connections := make(map[string]interface{})

	for _, command := range input {
		destination := strings.TrimSpace(strings.Split(command, "->")[1])
		instruction := strings.Split(strings.TrimSpace(strings.Split(command, "->")[0]), " ")

		// Convert instructions from []string to []interface{}
		new_instruction := make([]interface{}, len(instruction))
		for i, v := range instruction {
			new_instruction[i] = v
		}

		connections[destination] = new_instruction
	}

	ans := uint16(0)
	ans, connections = solve_connection(connections, "a")

	return ans
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
func part2(input []string) uint16 {
	b := part1(input)
	connections := make(map[string]interface{})

	for _, command := range input {
		destination := strings.TrimSpace(strings.Split(command, "->")[1])
		instruction := strings.Split(strings.TrimSpace(strings.Split(command, "->")[0]), " ")

		// Convert instructions from []string to []interface{}
		new_instruction := make([]interface{}, len(instruction))
		for i, v := range instruction {
			new_instruction[i] = v
		}

		connections[destination] = new_instruction
	}
	connections["b"] = b

	ans := uint16(0)
	ans, connections = solve_connection(connections, "a")

	return ans
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
func main() {
	// Read the arguments
	args := os.Args[1:]
	var fileName string
	if len(args) > 0 {
		fileName = args[0]
	} else {
		fileName = "input.txt"
	}
	// Start the timer!
	startTime := time.Now()

	// Read the input to an array of strings
	path := aoc_utils.GetFilePath(2015, 07, fileName)
	content, err := aoc_utils.ReadInputAsArray(path)
	if err != nil {
		fmt.Println("Error reading input:", err)
		return
	}

	// Run the Parts
	p1Result := aoc_utils.MeasurePerformance(func() interface{} {
		return part1(content)
	})

	p2Result := aoc_utils.MeasurePerformance(func() interface{} {
		return part2(content)
	})

	// End the Timer!
	endTime := time.Now()
	duration := endTime.Sub(startTime)

	resultsForDay := aoc_utils.AOCDayResults{
		Year:     2015,
		Day:      07,
		Part1:    p1Result,
		Part2:    p2Result,
		Duration: duration.Nanoseconds(),
		Timestamp: aoc_utils.AOCTimestamp{
			Start: startTime.Format(time.RFC3339),
			End:   endTime.Format(time.RFC3339),
		},
	}
	jsonResults, _ := json.Marshal(resultsForDay)
	fmt.Println(string(jsonResults))
}

File: events/2015/07/solution.rs
use aoc_utils::read_input_as_array;
use aoc_utils::measure_performance;
use aoc_utils::AOCDayResults;
use aoc_utils::AOCTimestamp;
use chrono::{DateTime, Utc};
use std::collections::HashMap;

#[derive(Debug, Clone)]
enum CommandValue {
    Int(u16),
    StringArray(Vec<String>),
}

type ConnectionRecord = HashMap<String, CommandValue>;

fn update_connections(mut collections: ConnectionRecord) -> ConnectionRecord {
  let keys_to_update: Vec<_> = collections
    .iter()
    .filter_map(|(wire, command)| {
      if matches!(command, CommandValue::Int(_)) {
        return None;
      }
      
      if let CommandValue::StringArray(arr) = command {
        if arr.len() == 1 {
          if let Ok(num) = arr[0].parse::<u16>() {
            return Some((wire.clone(), num));
          }
        }
      }
      None
    })
    .collect();

  for (wire, num) in keys_to_update {
    collections.insert(wire, CommandValue::Int(num));
  }

  collections
}

fn solve_connection(origin_connections: ConnectionRecord, to_solve: String) -> (u16, ConnectionRecord) {
  let mut connections = update_connections(origin_connections);
  let wc = connections.get(&to_solve);

  if let Some(CommandValue::Int(value)) = wc {
    return (*value, connections)
  }

  if let Some(CommandValue::StringArray(arr)) = connections.clone().get(&to_solve) {
    if arr.len() == 1 {
      let num: u16;
      let num_test = arr[0].parse::<u16>();
      match num_test {
        Ok(number) => num = number,
        Err(_) => return solve_connection(connections, arr[0].clone()),
      }
      connections.insert(to_solve, CommandValue::Int(num));
    } else if arr.len() == 2 {
      let nv = arr[1].clone();
      let mut ans: u16;

      let nv_test = nv.parse::<u16>();
      match nv_test {
        Ok(number) => ans = number,
        Err(_) => (ans, connections) = solve_connection(connections, nv),
      }

      ans = !ans;

      connections.insert(to_solve, CommandValue::Int(ans));
      return (ans, connections)
    } else if arr.len() == 3 {
      let op = arr[1].clone();

      let a: u16;
      let b: u16;

      let a_test = arr[0].parse::<u16>();
      match a_test {
        Ok(number) => a = number,
        Err(_) => (a, connections) = solve_connection(connections, arr[0].clone()),
      }

      let b_test = arr[2].parse::<u16>();
      match b_test {
        Ok(number) => b = number,
        Err(_) => (b, connections) = solve_connection(connections, arr[2].clone())
      }

      let ans: u16;
      let mut ans_set: bool = false;

      if op == "AND" {
        ans = a&b;
        ans_set = true;
      } else if op == "OR" {
        ans = a|b;
        ans_set = true;
      } else if op == "LSHIFT" {
        ans = a<<b;
        ans_set = true;
      } else if op == "RSHIFT" {
        ans = a>>b;
        ans_set = true;
      } else { ans = 0; }

      if ans_set {
        connections.insert(to_solve, CommandValue::Int(ans));
        return (ans, connections)
      }
    }
  }
  (0, connections)
}

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
fn part1(input: &Vec<String>) -> u16 {
  let mut connections: ConnectionRecord = HashMap::new();

  for command in input {
    let [raw_cmd, mut destination] = command
        .split("->")
        .collect::<Vec<&str>>()
        .try_into()
        .unwrap();
    destination = destination.trim();
    
    let cmd = raw_cmd
        .trim()
        .split(" ")
        .map(|s| s.to_string())
        .collect::<Vec<String>>();

    connections.insert(destination.to_string(), CommandValue::StringArray(cmd));
  }
  
  let (ans, _new_connections) = solve_connection(connections, "a".to_string());
  
  ans
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
fn part2(input: &Vec<String>) -> u16 {
  let b = part1(input);
  let mut connections: ConnectionRecord = HashMap::new();

  for command in input {
    let [raw_cmd, mut destination] = command
        .split("->")
        .collect::<Vec<&str>>()
        .try_into()
        .unwrap();
    destination = destination.trim();
    
    let cmd = raw_cmd
        .trim()
        .split(" ")
        .map(|s| s.to_string())
        .collect::<Vec<String>>();

    connections.insert(destination.to_string(), CommandValue::StringArray(cmd));
  }
  connections.insert("b".to_string(), CommandValue::Int(b));
  
  let (ans, _new_connections) = solve_connection(connections, "a".to_string());
  
  ans
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
fn main() {
  // Read the arguments
  let args = std::env::args().collect::<Vec<String>>();
  let default_file_name = String::from("input.txt");
  let file_name = args.get(1).unwrap_or(&default_file_name);

  // Start the timer
  let start_timestamp = std::time::SystemTime::now();
  let start_time = std::time::Instant::now();
  
  // Read the input
  let input = read_input_as_array(file_name).unwrap();

  // Run the Parts
  let p1_result = measure_performance(part1, &input);
  let p2_result = measure_performance(part2, &input);

  // Stop the timer
  let end_time = std::time::Instant::now();
  let end_timestamp = std::time::SystemTime::now();
  let duration = end_time.duration_since(start_time).as_nanos();

  let results = AOCDayResults {
    year: 2015,
    day: 07,
    part1: p1_result,
    part2: p2_result,
    duration: duration,
    timestamp: AOCTimestamp {
      start: DateTime::<Utc>::from(start_timestamp).to_rfc3339(),
      end: DateTime::<Utc>::from(end_timestamp).to_rfc3339(),
    },
  };

  let json_results = serde_json::to_string(&results).unwrap();
  println!("{}", json_results);
}